tests/test_traits.py

```diff
@@ -549,6 +561,440 @@ class SupeculatabilityTestOp(IRDLOperation):
+            [2, 4],
+            test.TestType("foo"),
+            [2, 3],
+        ),
+    ],
+)
+def test_same_operands_and_result_type_trait_for_result_element_type_of_shaped_types(
+    operand1_and_result_element_type: Attribute,
+    operand_and_result_shape1: tuple[int],
+    result_element_type2: Attribute,
+    result_shape2: tuple[int],
+):
+    op = SameOperandsAndResultTypeOp(
+        operands=[
+            TestSSAValue(
+                TensorType(operand1_and_result_element_type, operand_and_result_shape1)
+            )
+        ],
+        result_types=[
+            [
+                TensorType(operand1_and_result_element_type, operand_and_result_shape1),
+                TensorType(result_element_type2, result_shape2),
+            ],
+        ],
+    )
+
+    with pytest.raises(
+        VerifyException,
+        match="requires the same type for all operands and results",
+    ):
+        op.verify()
+
+
+@pytest.mark.parametrize(
+    "operands_num",
+    [1, 2, 3],
+)
+@pytest.mark.parametrize(
+    "results_num",
+    [1, 2, 3],
+)
+@pytest.mark.parametrize(
+    (
+        "operand_element_type",
+        "operand_shape",
+        "result_element_type",
+        "result_shape",
+    ),
+    [
+        (
+            test.TestType("foo"),
+            [2, 3],
+            test.TestType("qux"),
+            [2, 3],
+        ),
+        (
+            test.TestType("foo"),
+            [2, 3],
+            test.TestType("foo"),
+            [2, 4],
+        ),
+        (
+            test.TestType("qux"),
+            [2, 3],
+            test.TestType("foo"),
+            [2, 3],
+        ),
+        (
+            test.TestType("foo"),
+            [2, 4],
+            test.TestType("foo"),
+            [2, 3],
+        ),
+    ],
+)
+def test_same_operands_and_result_type_trait_for_element_type_of_shaped_types(
+    operand_element_type: Attribute,
+    operand_shape: tuple[int],
+    result_element_type: Attribute,
+    result_shape: tuple[int],
+    operands_num: int,
+    results_num: int,
+):
+    op = SameOperandsAndResultTypeOp(
+        operands=[
+            [
+                TestSSAValue(TensorType(operand_element_type, operand_shape)),
+            ]
+            * operands_num,
+        ],
+        result_types=[[TensorType(result_element_type, result_shape)] * results_num],
+    )
+
+    with pytest.raises(
+        VerifyException,
+        match="requires the same type for all operands and results",
+    ):
+        op.verify()
+
+    op = SameOperandsAndResultTypeOp(
+        operands=[
+            [
+                TestSSAValue(MemRefType(operand_element_type, operand_shape)),
+            ]
+            * operands_num,
+        ],
+        result_types=[[MemRefType(result_element_type, result_shape)] * results_num],
+    )
+
+    with pytest.raises(
+        VerifyException,
+        match="requires the same type for all operands and results",
+    ):
+        op.verify()
```

Comment:
these seem duplicated, unless I'm missing something

Reply:
One tests a `TensorType`, that could potentially have an encoding) and the other a `MemRefType` that doesn't have an encoding.
I'll add a comment.

Reply:
Ah yes I didn't see the difference, maybe could be a parametrized pytest?

Reply:
I'll give it a go

Reply:
I couldn't do it by adding a `parametrize` on just the container type (`TensorType`, `MemRefType`, etc.) and taking advantage of the Cartesian product with the ctor arguments because they are generics, so just denoting the class type as `TensorType[Any]` did not let me instantiate it in the test method.

Not sure if I'm missing something, but I couldn't find anything similar in the rest of the codebase.
The closest I found were the tests for bufferization and they just instantiate the type in the `parametrize` decorator.